blakserv/ccode.c

```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+      bprintf("C_PointInSector has bad col tag\n");
+      return NIL;
+   }
+
+   if (fr_val.v.tag != TAG_INT)
+   {
+      bprintf("C_PointInSector has bad fr tag\n");
+      return NIL;
+   }
+
+   if (id_list_val.v.tag == TAG_INT)
+   {
+      val_type temp, nil_val;
+      temp.v.tag = TAG_INT;
+      temp.v.data = id_list_val.v.data;
+
+      nil_val.v.tag = TAG_INT;
+      nil_val.v.data = 0;
+
+      id_list_val.v.data = Cons(temp, nil_val);
+      id_list_val.v.tag = TAG_LIST;
+   }
+   else if (id_list_val.v.tag != TAG_LIST)
+   {
+      bprintf("C_PointInSector has bad 6th parameter tag, expected list or int\n");
+      return NIL;
+   }
+
+   roomdata_node *rd = GetRoomDataByID(room_val.v.data);
+   if (!rd || !rd->file_info.sectors)
+   {
+      bprintf("C_PointInSector has bad room id passed in: %i\n", room_val.v.data);
+      return NIL;
+   }
+
+   room_type *r = &rd->file_info;
+
+   // remember that kod uses 1-based arrays and we don't
+   int row0 = (int) row_val.v.data - 1;
+   int col0 = (int) col_val.v.data - 1;
+
+   // Now, we need to convert the kod coordinates to coordinates that match the room data
+   int fine_x = (int) ((long long) (col0 * FINENESS) + fc_val.v.data) * POLYGON_COORD_SCALE;
+   int fine_y = (int) ((long long) (row0 * FINENESS) + fr_val.v.data) * POLYGON_COORD_SCALE;
+
+   // Build the vector of sector ids
+   if (id_list_val.v.tag == TAG_INT)
+   {
+      sector_id = (int) id_list_val.v.data;
+      sector_ids.push_back(sector_id);
+   }
+   else if (id_list_val.v.tag == TAG_LIST)
+   {
+      list_node *list = GetListNodeByID(id_list_val.v.data);
+      if (!list)
+      {
+         bprintf("C_PointInSector failed to get list node by id: %i\n", id_list_val.v.data);
+         return NIL;
+      }
+
+      while (list != NULL)
+      {
+         if (list->first.v.tag == TAG_INT)
+         {
+            sector_id = list->first.v.data;
+            sector_ids.push_back(sector_id);
+         }
+         else
+         {
+            bprintf("C_PointInSector: Sector list element is not an integer\n");
+         }
+
+         if (list->rest.v.tag == TAG_LIST)
+         {
+            list = GetListNodeByID(list->rest.v.data);
+         }
+         else
+         {
+            break;
+         }
+      }
+   }
+   else
+   {
+      bprintf("C_PointInSector has bad 6th parameter tag, expected list or int\n");
+      return NIL;
+   }
+
+   for (int sector_id : sector_ids)
+   {
+      for (int i = 0; i < r->num_sectors; i++)
+      {
+         if (r->sectors[i].id != sector_id)
+         {
+            continue;
+         }
+
+         // Found matching sector, check polygons
+         for (int p = 0; p < r->sectors[i].num_polygons; p++)
+         {
+            // First, check if the sector has polygons
+            if (r->sectors[i].polygons == NULL || p >= r->sectors[i].num_polygons)
+            {
+               continue;
+            }
+
+            server_polygon *poly = &r->sectors[i].polygons[p];

```

Comment:
This is a good place to use a reference:

const server_polygon &poly = r->sectors[i].polygons[p];